snuba/util.py

```diff
@@ -325,3 +329,21 @@ def create_metrics(prefix: str, tags: Optional[Tags] = None) -> MetricsBackend:
             else None,
         ),
     )
+
+
+def with_span(op="function", var=None):
+    def decorator(func):
+        frame_info = inspect.stack()[1]
+        filename = frame_info.filename
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            with sentry_sdk.start_span(description=func.__name__, op=op) as span:
+                span.set_data("filename", filename)
+                if var:
+                    kwargs[var] = span
+                return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
```

Comment:
> I'm assuming the function decorator will be contentious, though.

I can't speak for everyone else, but I personally don't have any specific philosophical issues with using a decorator for this — though it seems a bit limited in flexibility since it requires anything being wrapped to be a callable which might be unnecessary some cases. I was going to suggest using a [`ContextDecorator`](https://docs.python.org/3/library/contextlib.html#using-a-context-manager-as-a-function-decorator) but at that point it seems like this would be pretty close to the original SDK API.

I think it's important that these functions maintain their type signatures, and typing this effectively is going to be a challenge. Typing any decorator is kind of a pain, especially ones that change the call arguments to the inner function. mypy doesn't do a great job distinguishing between positional and keyword arguments, so I'm not sure you'll be able to find an effective strategy here. There is [some extended `Callable`](https://mypy.readthedocs.io/en/stable/additional_features.html#extended-callable-types) support but going down that path seems like it has a smell of excessive cleverness. In general, I think we should try and avoid things like keyword-only arguments when possible (for multiple reasons but typing support is probably the least subjective one.)

Reply:
Yeah, if I created some sort of context I may as well just use the context from the SDK, probably.

I'll dump the capability for forwarding the span and use the scope context functionality from the SDK.

Reply:
Done.